ngraph_bridge/ngraph_builder.cc

```diff
@@ -1898,28 +1898,14 @@ static Status TranslateLog1pOp(
 static Status TranslateLogSoftmaxOp(const Node* op,
                                     const std::vector<const Tensor*>&,
                                     Builder::OpMap& ng_op_map) {
-  ng::Output<ng::Node> ng_inp;
-  TF_RETURN_IF_ERROR(GetInputNodes(ng_op_map, op, ng_inp));
-  auto inp_shape = ng_inp.get_shape();
-  size_t rank = inp_shape.size();
-  auto ng_axis = ng::AxisSet{rank - 1};
-  // Batch i, class j
-  // logsoftmax[i, j] = logits[i, j] - log(sum(exp(logits[i])))
-  // Actually implementing: logsoftmax[i, j] = logits[i, j] - max(logits[i]) -
-  // log(sum(exp(logits[i] - max(logits[i]))))
-  auto ng_max = ConstructNgNode<ng::op::Broadcast>(
-      op->name(), ConstructNgNode<ng::op::Max>(op->name(), ng_inp, ng_axis),
-      inp_shape, ng_axis);
-  auto ng_inp_minus_max =
-      ConstructNgNode<opset::Subtract>(op->name(), ng_inp, ng_max);
-  auto ng_exp = ConstructNgNode<ng::op::Exp>(op->name(), ng_inp_minus_max);
-  auto ng_log_sum = ConstructNgNode<ng::op::Log>(
-      op->name(), ConstructNgNode<ng::op::Sum>(op->name(), ng_exp, ng_axis));
-  auto ng_broadcast = ConstructNgNode<ng::op::Broadcast>(
-      op->name(), ng_log_sum, ng_inp.get_shape(), ng_axis);
-  auto ng_output = ConstructNgNode<opset::Subtract>(
-      op->name(), ng_inp_minus_max, ng_broadcast);
-  SaveNgOp(ng_op_map, op->name(), ng_output);
+  ng::Output<ng::Node> ng_input;
+  TF_RETURN_IF_ERROR(GetInputNodes(ng_op_map, op, ng_input));
+  auto input_shape = ng_input.get_shape();
+  auto rank = input_shape.size();
+  auto ng_log_softmax = ConstructNgNode<opset::Log>(
+      op->name(),
+      ConstructNgNode<opset::Softmax>(op->name(), ng_input, rank - 1));
```

Comment:
Softmax isn't numerically stable especially against under- and overflow; you need to normalize the input here (`logits - max(logits)`) to avoid stability errors.

See https://stackoverflow.com/a/52132033.

Reply:
Yes, foir numerical stability we should calculate like: logsoftmax[i, j] = logits[i, j] - max(logits[i]) -	  SaveNgOp(ng_op_map, op->name(), ng_log_softmax);
  // log(sum(exp(logits[i] - max(logits[i]))))
where, we use  a multiplier C such that log(C) = - max(logits)